File: day19/day19.py
#!/usr/bin/env python3

import sys

ADDR = 0
ADDI = 1
MULR = 2
MULI = 3
BANR = 4
BANI = 5
BORR = 6
BORI = 7
SETR = 8
SETI = 9
GTIR = 10
GTRI = 11
GTRR = 12
EQIR = 13
EQRI = 14
EQRR = 15
OPCOUNT = 16

NAMES = {
    "addr": ADDR,
    "addi": ADDI,
    "mulr": MULR,
    "muli": MULI,
    "banr": BANR,
    "bani": BANI,
    "borr": BORR,
    "bori": BORI,
    "setr": SETR,
    "seti": SETI,
    "gtir": GTIR,
    "gtri": GTRI,
    "gtrr": GTRR,
    "eqir": EQIR,
    "eqri": EQRI,
    "eqrr": EQRR,
}

def execute(op, a, b, c, regs):
    if op == ADDR: regs[c] = regs[a] + regs[b]
    elif op == ADDI: regs[c] = regs[a] + b
    elif op == MULR: regs[c] = regs[a] * regs[b]
    elif op == MULI: regs[c] = regs[a] * b
    elif op == BANR: regs[c] = regs[a] & regs[b]
    elif op == BANI: regs[c] = regs[a] & b
    elif op == BORR: regs[c] = regs[a] | regs[b]
    elif op == BORI: regs[c] = regs[a] | b
    elif op == SETR: regs[c] = regs[a]
    elif op == SETI: regs[c] = a
    elif op == GTIR: regs[c] = (a > regs[b]) and 1 or 0
    elif op == GTRI: regs[c] = (regs[a] > b) and 1 or 0
    elif op == GTRR: regs[c] = (regs[a] > regs[b]) and 1 or 0
    elif op == EQIR: regs[c] = (a == regs[b]) and 1 or 0
    elif op == EQRI: regs[c] = (regs[a] == b) and 1 or 0
    elif op == EQRR: regs[c] = (regs[a] == regs[b]) and 1 or 0
    else:
        assert 0, "Unknown operand {}".format(op)

class System(object):
    def __init__(self, txt):
        self.regs = [0] * 6
        self.program = []
        for line in txt.splitlines():
            lst = line.split(" ")
            if lst[0] == "#ip":
                self.ip = int(lst[1])
            else:
                op = NAMES[lst[0]]
                a, b, c = map(int, lst[1:])
                self.program.append((op, a, b, c))

    def reset(self):
        self.regs = [0] * 6

    def run(self, boost = False):
        while 0 <= self.regs[self.ip] < len(self.program):
            if boost and self.regs[self.ip] == 1:
                # in regs[0] there is sum of its divisors
                reg = self.program[self.regs[self.ip]+3][2]
                value = self.regs[reg]
                result = value + 1
                for i in range(2, value//2 + 1):
                    if value % i == 0:
                        result += i

                return result
            else:
                op, a, b, c = self.program[self.regs[self.ip]]
                execute(op, a, b, c, self.regs)
                self.regs[self.ip] += 1

        return self.regs[0]

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    s = System(txt)
    s.reset()
    print("Part1:", s.run(True))
    s.reset()
    s.regs[0] = 1
    print("Part2:", s.run(True))

File: day19/day19.c
#include <assert.h>
#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

enum
{
	OP_ADDI,
	OP_ADDR,
	OP_BANI,
	OP_BANR,
	OP_BORI,
	OP_BORR,
	OP_EQIR,
	OP_EQRI,
	OP_EQRR,
	OP_GTIR,
	OP_GTRI,
	OP_GTRR,
	OP_MULI,
	OP_MULR,
	OP_SETI,
	OP_SETR,
	OP_COUNT,
};

static const char *symbols[] = {
	"addi",
	"addr",
	"bani",
	"banr",
	"bori",
	"borr",
	"eqir",
	"eqri",
	"eqrr",
	"gtir",
	"gtri",
	"gtrr",
	"muli",
	"mulr",
	"seti",
	"setr",
};

static int find_opcode(const char *symbol)
{
	int low = 0;
	int high = OP_COUNT-1;
	while (low < high)
	{
		int mid = low + (high - low)/2;
		if (strcmp(symbols[mid], symbol) < 0)
		{
			low = mid + 1;
		}
		else
		{
			high = mid;
		}
	}

	if (strcmp(symbols[high], symbol) == 0)
	{
		return high;
	}

	return -1;
}

static void execute(int64_t *reg, int opcode, int a, int b, int c)
{
	switch (opcode)
	{
	case OP_ADDR: reg[c] = reg[a] + reg[b]; break;
	case OP_ADDI: reg[c] = reg[a] + b; break;
	case OP_MULR: reg[c] = reg[a] * reg[b]; break;
	case OP_MULI: reg[c] = reg[a] * b; break;
	case OP_BANR: reg[c] = reg[a] & reg[b]; break;
	case OP_BANI: reg[c] = reg[a] & b; break;
	case OP_BORR: reg[c] = reg[a] | reg[b]; break;
	case OP_BORI: reg[c] = reg[a] | b; break;
	case OP_SETR: reg[c] = reg[a]; break;
	case OP_SETI: reg[c] = a; break;
	case OP_GTIR: reg[c] = a > reg[b]; break;
	case OP_GTRI: reg[c] = reg[a] > b; break;
	case OP_GTRR: reg[c] = reg[a] > reg[b]; break;
	case OP_EQIR: reg[c] = a == reg[b]; break;
	case OP_EQRI: reg[c] = reg[a] == b; break;
	case OP_EQRR: reg[c] = reg[a] == reg[b]; break;
	default: assert(0); break;
	}
}

struct cpu
{
	int64_t regs[6];
	int boundreg;
};

struct program
{
	struct instruction
	{
		int opcode;
		int a;
		int b;
		int c;
	} *instr;
	size_t count;
	size_t size;
};

static void cpu_reset(struct cpu *c)
{
	int boundreg = c->boundreg;
	memset(c, 0, sizeof(*c));
	c->boundreg = boundreg;
}

static int64_t cpu_execute(struct cpu *c, struct program *p, int boost)
{
	size_t ip = 0;
	while (ip < p->count)
	{
		struct instruction *ins = p->instr + ip;
		if (boost && ip == 1)
		{
			int r = p->instr[ip + 3].b;
			int64_t value = c->regs[r];
			int64_t result = value + 1;
			for (int64_t i = 2, e = value / 2 + 1; i < e; i++)
			{
				if (value % i == 0)
				{
					result += i;
				}
			}
			return result;
		}
		else
		{
			execute(c->regs, ins->opcode, ins->a, ins->b, ins->c);
		}
		ip = ++c->regs[c->boundreg];
	}
	return c->regs[0];
}

static void program_free(struct program *p)
{
	if (p)
	{
		free(p->instr);
		free(p);
	}
}

static struct program *program_load(FILE *input)
{
	struct program *p = calloc(1, sizeof(*p));
	if (!p)
	{
		return NULL;
	}

	char *line = NULL;
	size_t linesize = 0;
	while (getline(&line, &linesize, input) != -1)
	{
		if (p->count == p->size)
		{
			size_t newsize = p->size ? p->size * 2 : 2;
			struct instruction *newi = realloc(
				p->instr, newsize * sizeof(p->instr[0]));
			if (!newi)
			{
				program_free(p);
				return NULL;
			}
			p->size = newsize;
			p->instr = newi;
		}

		char *space = strchr(line, ' ');
		if (!space)
		{
			continue;
		}

		*space = 0;
		p->instr[p->count].opcode = find_opcode(line);
		sscanf(space+1, "%d %d %d",
		       &p->instr[p->count].a,
		       &p->instr[p->count].b,
		       &p->instr[p->count].c);
		p->count++;
	}
	free(line);

	return p;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fprintf(stderr, "Usage: %s <filename>\n", argv[0]);
		return 1;
	}

	FILE *input = fopen(argv[1], "rb");
	if (!input)
	{
		fprintf(stderr, "Cannot open %s\n", argv[1]);
		return 1;
	}

	struct cpu cpu = {0};
	fscanf(input, "#ip %d", &cpu.boundreg);
	struct program *p = program_load(input);
	fclose(input);
	if (!p)
	{
		fprintf(stderr, "Cannot parse the data\n");
		return 1;
	}

	printf("Part1: %" PRId64 "\n", cpu_execute(&cpu, p, 1));

	cpu_reset(&cpu);
	cpu.regs[0] = 1;
	printf("Part2: %" PRId64 "\n", cpu_execute(&cpu, p, 1));
	program_free(p);

	return 0;
}
